### ISSUE TITLE
Peer QR Code Generator

### ISSUE THREAD

**Issue by theonemcdonald** [theonemcdonald/pfSense-pkg-WireGuard#7]
Wishlist:

The ability to click a shortcut on a peer and have a scannable QR code that can be scanned directly into the WireGuard applications

**Comment by theonemcdonald** [comment]
![screen5](https://user-images.githubusercontent.com/3102039/114230582-51db8200-9947-11eb-86e6-e40dfb6f594d.PNG)

**Comment by vajonam** [comment]
This is neat way to do it, but depends on qrencode - https://github.com/linuxserver/docker-wireguard/blob/master/root/app/show-peer

**Comment by theonemcdonald** [comment]
libqrencode is available in the FreeBSD ports tree :)

**Comment by vajonam** [comment]
I suspect we have create the peer private keys which is different from now and the send that over a QR. or config file.

**Comment by theonemcdonald** [comment]
Yea there will have to be some flow changes for sure. I'm open to suggestions! Currently I'm thinking about adding an 'exportable' checkbox in the peer configuration pop-up that will change the public key field to a private key field. It will also show a dropdown with peer export profiles. These profiles can be created on a new tab. These profiles will contain things like defining the public endpoint FQDN/IP and port, DNS servers and search domain, etc.

**Comment by vajonam** [comment]
Yeah I like it. works for me. maybe bit confusing on the "exportable", but I understand why we need it. I like the QR/download icon near the edit pencil.  

maybe a separate section called exprotable (when checkbox is ticked) . which means. all the export stuff shows up in once place, private key, export profile etc. then it hides the pubkey and shows the privkey on the export section.

**Comment by mtoczycki** [comment]
What about using email to send qr codes and private keys, to the user?  I don’t have a whole lot of experience with WireGuard but emailing the keys would be great as well.

**Comment by Tigger2014** [comment]
I think sending keys via email is a bit insecure without an additional layer of security on top. Might be something thats possible though

**Comment by 4o66** [comment]
> I think sending keys via email is a bit insecure without an additional layer of security on top. Might be something thats possible though

Newbie to pfSense, but would it be possible to add the option to tie a profile to the local user database, then require them to log into the pfSense webGUI, navigate to VPN then display profile? 

What sort of login rights do non admin users have on pfSense?

**Comment by theonemcdonald** [comment]
@4o66 That might actually be possible. Sort of like a self-service portal

**Comment by GenericStudent** [comment]
Hello all!

Loving this WireGuard plugin so thought I'd throw a couple options into this feature request.

Would like to preface it with the fact that FreeBSD development is new to me, as is PHP ( been a long time since I was developing using PHP so the code quality may be aweful etc :) ).

Quite a few TODOs still left with my implementation:
- Cannot for the life of me get the dependencies via the Makefile to actually install graphics/libqrencode, trying it manually on SSH CLI yields no better results, so in order for me to actually get QR codes working I had to SCP the pkg's onto my pfSense and install
- Massive assertions on the flow of things made by me, such as adding a Private key readonly input to the Peers Edit page which allows you to generate private keys (but not store them in config), my thinking here is if we're installing the tunnel via QR Code, the admin adding the peer probably doesn't have an existing private / public key to add to the peer (and it's required to create a peer), so added this capability just to get around the public key requirement
- Couldn't get a confirmation on the usepost anchor tag for QR Code generation from Peer list page, this results in the link using POST but not warning the user before it overwrites the peer's public key
- Not 100% sure the "apply changes" part of the QR Code generation section is perfect
- Am not sure on how we set the pfsense peer's endpoint in the generated QR Code as the tunnel doesn't seem to have an option for it, would we look up the associated interface? But then how'd we get a public FQDN for this? Leaving the exported config seem to leave it down to the user to add it when importing onto mobile, would be lovely if we could set this for them though!
- Code refactors to make it line up better with existing buckets of code in the wg_*.inc scripts

The new QR code generation button on Peer List / index page, this will overwrite the peer's public key and export the private key into the QR code seen when clicked, this means it's not persisted anywhere which is good :), QR Code button is only available if the peer is assigned to a tunnel
![peers_list](https://user-images.githubusercontent.com/13488006/123541913-b46e3e00-d73e-11eb-8405-bdc8e4fe2d44.png)

When the generate QR code button is pressed, the user can then scan this QR code into their WireGuard mobile app, size was increased to make it scan more reliably at ~30cm from screen
![peers_list_qrcode_clicked](https://user-images.githubusercontent.com/13488006/123541914-b506d480-d73e-11eb-9543-c8d0a47d3623.png)

Peers Edit page with the new readonly private key field
![peers_edit](https://user-images.githubusercontent.com/13488006/123541917-b6380180-d73e-11eb-8239-3e1273c77b18.png)

Android WireGuard app view when the imported config (only had to give it a name after scanning, seems you cannot specify name (shame as we have peer description AND tunnel description at hand))
![wireguard-android-imported-view](https://user-images.githubusercontent.com/13488006/123541916-b59f6b00-d73e-11eb-84cc-00368d2db67e.png)

Viewing the edit peer page on WireGuard Android app
![wireguard-android-imported-edit](https://user-images.githubusercontent.com/13488006/123541915-b59f6b00-d73e-11eb-946e-150631a843e3.png)

Would love any comments / feedback. i suspect this may be an acceptable short term workflow and a longer term, more self-service style workflow may be more preferable in the long term / the better design choices.

**Comment by Tigger2014** [comment]
@GenericStudent 

Nice work on this. I can't add much to help as php is not my area.

However regarding installing dependencies i think there was talk about using a pure js qr generator. There are quite a few examples around so that might be worth a look.

**Comment by theonemcdonald** [comment]
@GenericStudent 

Thanks for the code! I like where this is going. There are a few things on my internal roadmap that will likely supercede some of what you're doing (like peer export profiles) that will make QR and .conf export more sensible. One thing worth mentioning is that because a public key can be trivially derived from a private key, it isn't really useful to store both a private and public key pair. The only reason why both are stored in the XML for tunnels is because this project originally forked from Netgate's UI in 2.5.0 and they were storing the entire keypair.

My thinking is peers will have a new flag "Exportable" that when ticked, the "key" field will be treated like a private key... otherwise it will be treated like a public key. They way we aren't storing anything more than what we absolutely have to. And in fact, most cases you wouldn't know the private key of the peer anyways...well at least you shouldn't, and thus it wouldn't be exportable.

Peer export profiles would contain the requisite extra bits to actually generate a useful conf, like public endpoint , port, DNS settings, etc.

**Comment by GenericStudent** [comment]
@theonemcdonald that all sounds really useful, would you envisage export profiles having a hierarchy? Such that a tunnel can have an endpoint, dns, and mtu values that are shared between all peers, but peers can then override each if required.

The current implementation doesn't actually store the private key, only the public key it derives. 

Do you have the peer export profiles on a feature branch / in progress? Happy to help anywhere I can.

Adam
